In [11]:
import kaggle
import zipfile
import pandas as pd

In [12]:
# downloading a dataset
!kaggle datasets download -d rashikrahmanpritom/data-science-job-posting-on-glassdoor

Dataset URL: https://www.kaggle.com/datasets/rashikrahmanpritom/data-science-job-posting-on-glassdoor
License(s): CC0-1.0




  0%|          | 0.00/1.53M [00:00<?, ?B/s]
 65%|██████▌   | 1.00M/1.53M [00:00<00:00, 2.06MB/s]
100%|██████████| 1.53M/1.53M [00:00<00:00, 2.52MB/s]
100%|██████████| 1.53M/1.53M [00:00<00:00, 2.41MB/s]


In [13]:
# extracting a zip file
zipfile_name = 'data-science-job-posting-on-glassdoor.zip'
with zipfile.ZipFile(zipfile_name, 'r') as file:
    file.extractall()

*note*: at this oint, we have two files:

- Uncleaned_DS_jobs.csv
- Cleaned_DS_Jobs.csv

let's **get rid of** Cleaned_DS_Jobs.csv now

In [14]:
import os

In [15]:
file_to_rm = 'Cleaned_DS_Jobs.csv'
os.remove(file_to_rm)
# changing a name of a file
os.rename('Uncleaned_DS_jobs.csv', 'job_posting.csv')

In [16]:
# reading csv file as pandas dataframe
job_posting = pd.read_csv('job_posting.csv')

In [17]:
# let's checkout our dataframe
job_posting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 672 entries, 0 to 671
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   index              672 non-null    int64  
 1   Job Title          672 non-null    object 
 2   Salary Estimate    672 non-null    object 
 3   Job Description    672 non-null    object 
 4   Rating             672 non-null    float64
 5   Company Name       672 non-null    object 
 6   Location           672 non-null    object 
 7   Headquarters       672 non-null    object 
 8   Size               672 non-null    object 
 9   Founded            672 non-null    int64  
 10  Type of ownership  672 non-null    object 
 11  Industry           672 non-null    object 
 12  Sector             672 non-null    object 
 13  Revenue            672 non-null    object 
 14  Competitors        672 non-null    object 
dtypes: float64(1), int64(2), object(12)
memory usage: 78.9+ KB


so, ``job_posting`` dataframe consists of:
- 672 rows
- 15 columns

In [18]:
# how our uncleaned dataset looks like
job_posting.head()

index          Job Title               Salary Estimate  \
0      0  Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1      1     Data Scientist  $137K-$171K (Glassdoor est.)   
2      2     Data Scientist  $137K-$171K (Glassdoor est.)   
3      3     Data Scientist  $137K-$171K (Glassdoor est.)   
4      4     Data Scientist  $137K-$171K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  
0            EmblemHealth, UnitedHealth Group, Aetna  
1                                                 -1  
2                                                 -1  
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4               Commerce Signals, Cardlytics, Yodlee

right off the bat, there are several changes we can do to make our data more presentable:
1. in *Salary Estimate* column: get rid of dollar signs and *(Glassdoor est.)*
2. in *Company Name* column: left slice everything before newline character appearance
3. in *Job Description* column: get rid of newline chars, maybe replace them with ' '(?)
4. in a whole dataframe: replace *-1* and *Unknown / Non-Applicable* values to *Unknown*, since both of those values that we want to replace basically mean the same thing
5. add the following columns based on *Salary Estimate*:
    - *Min salary*
    - *Max salary*
    - *Average salary*
6. add *Same state* column, that will represent whether or not a companys location and headquartes are located in the same state

#### removing chars

In [43]:
# removing ' (Glassdoor est.)'
job_posting['Salary Estimate'] = job_posting['Salary Estimate'].map(lambda x: x.split('('))

salary_est = job_posting['Salary Estimate']

def extract_salary(est):
    min_salary, max_salary = est.split('-')

    min_salary = int(min_salary[1:-1]) * 1000
    max_salary = max_salary[1:-1]

    return min_salary, max_salary

# job_posting[['Min Salary', 'Max Salary']] = job_posting['Salary Estimate'].apply(lambda x: pd.Series(extract_salary(x)))
# job_posting['Max Salary'].unique()


AttributeError: 'list' object has no attribute 'split'